In [6]:
# to load the environment variables from a .env file
from dotenv import load_dotenv
import os
load_dotenv()

True

In [7]:
# to get the Groq API key from environment variables
groq_api_key = os.getenv("GROQ_API_KEY")
# to initialize llm with Groq
from langchain_groq import ChatGroq
llm = ChatGroq(api_key=groq_api_key, model="llama-3.3-70b-versatile")
llm

ChatGroq(client=<groq.resources.chat.completions.Completions object at 0x000001A1AD99C0D0>, async_client=<groq.resources.chat.completions.AsyncCompletions object at 0x000001A1AD793990>, model_name='llama-3.3-70b-versatile', model_kwargs={}, groq_api_key=SecretStr('**********'))

In [15]:
# import necessary modules
from langchain.schema import (
    AIMessage,
    HumanMessage,
    SystemMessage   
)

In [17]:
speech = """
My dear citizens of India,
Today, I stand before you with pride and hope as our nation steps into a new era of progress. India has always been a land of courage, culture, and intelligence. From the ancient days of Nalanda and Takshashila to our modern space missions, we have consistently shown the world the power of Indian determination.

Our government believes that development should reach every corner of India. Whether a child is born in a small village in Odisha or in a big city like Mumbai, the opportunity for growth must be the same. Our priority is to build strong foundations — better roads, clean drinking water, more hospitals, smart schools, and reliable electricity in every home.

In the last few years, technology has become India’s biggest partner. With digital services, citizens no longer need to stand in long lines for essential tasks. Aadhaar, UPI payments, online education, and e-governance systems are helping people live easier, faster, and more secure lives. This is not just modernization; it is a movement towards equality.

We are also supporting our youth — the strength of our nation. Programs focusing on skill development, startups, and innovation are helping young Indians convert their ideas into successful careers. Our vision is not just to create jobs, but to create employers who will take India to global leadership in industries like artificial intelligence, space technology, and renewable energy.

Agriculture remains the backbone of our nation. We are bringing more technology to farms, ensuring farmers get a fair price for their produce and better access to markets. Our goal is to transform farming into a profitable, respected profession, not a struggle for survival.

India’s role in global peace and climate protection is stronger than ever. We believe that sustainable development is the only path forward. New investments in green energy, electric mobility, and forest conservation will protect our environment for future generations. Our earth is a shared home, and India will always be a responsible guardian of nature.

Unity is our greatest strength. India is a country of many languages, religions, cultures, and traditions — yet we stand together as one. Discrimination has no place in the India we are building. Every citizen, regardless of background, deserves dignity and opportunity.

As we move toward the vision of a Developed India by 2047, every small step counts. The contribution of a teacher, the creativity of a student, the hard work of a farmer, the service of a soldier — all these form the soul of our progress. Let us promise to work together, think big, and keep India first in everything we do.

My dear friends, the future is full of opportunities. Let us embrace change, innovate with courage, and lead the world with confidence. The India of tomorrow is not just a dream — it is a destination we will reach together.
Jai Hind! Jai Bharat!
"""

In [18]:
# define the messages for the chat model
chat_messages = [
    SystemMessage(content="You are a helpful assistant that summarizes speeches."),
    HumanMessage(content=f"Summarize the following speech in concise points:\n\n{speech}")
]

In [19]:
# to get the how many tokens are used
llm.get_num_tokens(speech)  # integer value, to get we have to pass string input

598

In [20]:
# one way of getting the response from the llm -- first way of text summerization
response = llm(chat_messages)# AIMessage object
response.content

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_1740\89935592.py:2: LangChainDeprecationWarning: The method `BaseChatModel.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = llm(chat_messages)# AIMessage object


"Here are the concise points summarizing the speech:\n\n* India is entering a new era of progress, driven by courage, culture, and intelligence.\n* The government's priorities include:\n\t+ Building strong foundations (roads, water, hospitals, schools, electricity)\n\t+ Ensuring equal opportunities for growth across the country\n* Technology is a key partner in India's development, with digital services and innovations like Aadhaar, UPI, and e-governance.\n* The government is supporting youth through skill development, startups, and innovation programs.\n* Agriculture is being transformed into a profitable profession through technology and better market access.\n* India is committed to global peace, climate protection, and sustainable development, with investments in green energy and conservation.\n* Unity and diversity are India's strengths, with a commitment to dignity and opportunity for all citizens.\n* The vision is to achieve a Developed India by 2047, with every citizen's contri

## Prompt Template Text Summerization

In [21]:
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain

# define the prompt template for text summerization
generic_template = """
Write the summary of the following speech in concise points:
{speech}
Translate the precise summary to {language}.
"""
prompt = PromptTemplate(
    input_variables=["speech", "language"],
    template=generic_template
)
prompt

PromptTemplate(input_variables=['language', 'speech'], input_types={}, partial_variables={}, template='\nWrite the summary of the following speech in concise points:\n{speech}\nTranslate the precise summary to {language}.\n')

In [22]:
# to form the complete prompt
final_prompt = prompt.format(speech=speech, language="Kannada")
final_prompt

'\nWrite the summary of the following speech in concise points:\n\nMy dear citizens of India,\nToday, I stand before you with pride and hope as our nation steps into a new era of progress. India has always been a land of courage, culture, and intelligence. From the ancient days of Nalanda and Takshashila to our modern space missions, we have consistently shown the world the power of Indian determination.\n\nOur government believes that development should reach every corner of India. Whether a child is born in a small village in Odisha or in a big city like Mumbai, the opportunity for growth must be the same. Our priority is to build strong foundations — better roads, clean drinking water, more hospitals, smart schools, and reliable electricity in every home.\n\nIn the last few years, technology has become India’s biggest partner. With digital services, citizens no longer need to stand in long lines for essential tasks. Aadhaar, UPI payments, online education, and e-governance systems a

In [23]:
# to get the tokens of the final prompt
llm.get_num_tokens(final_prompt)  # integer value, to get we have to pass string input

623

initial speech contain 598, but final prompt conain 621 because i added some more texts to speech which is " generic_template"

In [24]:
# to create a llm chain for text summerization using prompt template
chain = LLMChain(llm=llm, prompt=prompt)
summary = chain.run(speech=speech, language="Kannada")
summary

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_1740\1033343791.py:2: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use :meth:`~RunnableSequence, e.g., `prompt | llm`` instead.
  chain = LLMChain(llm=llm, prompt=prompt)
C:\Users\ADMIN\AppData\Local\Temp\ipykernel_1740\1033343791.py:3: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  summary = chain.run(speech=speech, language="Kannada")


"**English Summary:**\n\n* India is entering a new era of progress, driven by courage, culture, and intelligence.\n* The government aims to provide equal opportunities for growth across the country, with a focus on building strong foundations such as infrastructure, healthcare, and education.\n* Technology has become a key partner in India's development, enabling digital services, online education, and e-governance.\n* The government is supporting youth through skill development, startups, and innovation, aiming to create employers and leaders in industries like AI, space technology, and renewable energy.\n* Agriculture is being transformed into a profitable profession through technology and better market access.\n* India is committed to sustainable development, global peace, and climate protection, with investments in green energy, electric mobility, and forest conservation.\n* Unity and diversity are India's greatest strengths, with a focus on dignity and opportunity for all citizens

## StuffDocumentChain Text Summerization

In [2]:
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader("drapj.pdf")
docs = loader.load_and_split()
docs

d:\GenAI\LangChain\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


[Document(metadata={'producer': 'Skia/PDF m127', 'creator': 'Chromium', 'creationdate': '2025-10-27T13:45:14+00:00', 'moddate': '2025-10-27T13:45:14+00:00', 'source': 'drapj.pdf', 'total_pages': 2, 'page': 0, 'page_label': '1'}, page_content='Here is the full text of Dr. APJ Abdul Kalam\'s "My Vision for India" speech, delivered at IIT\nHyderabad on 25 May 2011, without any summarization:\nI have three visions for India. In 3000 years of our history people from all over the world have\ncome and invaded us, captured our lands, conquered our minds. From Alexander onwards the\nGreeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them\ncame and looted us, took over what was ours. Yet we have not done this to any other nation. We\nhave not conquered anyone. We have not grabbed their land, their culture and their history and\ntried to enforce our way of life on them. Why? Because we respect the freedom of others. That is\nwhy my FIRST VISION is that of FR

In [4]:
# Prompt template for text summerization using StuffDocumentChain
from langchain.prompts import PromptTemplate
prompt_template = """
Write short summary of the following speech in concise points,
speech: {text}
"""
prompt = PromptTemplate(
    input_variables=["text"],
    template=prompt_template
)

In [8]:
from langchain.chains.summarize import load_summarize_chain
chain = load_summarize_chain(llm, chain_type="stuff", prompt=prompt,verbose=True)# first it will combine all the documents into one and then will pass prompt to llm
summary = chain.run(docs)
summary

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_18032\2873647029.py:3: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  summary = chain.run(docs)




> Entering new StuffDocumentsChain chain...


> Entering new LLMChain chain...
Prompt after formatting:

Write short summary of the following speech in concise points,
speech: Here is the full text of Dr. APJ Abdul Kalam's "My Vision for India" speech, delivered at IIT
Hyderabad on 25 May 2011, without any summarization:
I have three visions for India. In 3000 years of our history people from all over the world have
come and invaded us, captured our lands, conquered our minds. From Alexander onwards the
Greeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them
came and looted us, took over what was ours. Yet we have not done this to any other nation. We
have not conquered anyone. We have not grabbed their land, their culture and their history and
tried to enforce our way of life on them. Why? Because we respect the freedom of others. That is
why my FIRST VISION is that of FREEDOM. I believe that India got its first vision of this in 1857,
when we 

'Here is the speech:\n Dr. A.P.J Abdul Kalam: Here is the full text of the speech "My Vision for India" delivered by Dr.\nAPJ Abdul Kalam at IIT Hyderabad on 25 May 2011.\nI have three visions for India. In 3000 years of our history people from all over the world have\ncome and invaded us, captured our lands, conquered our minds. From Alexander onwards the\nGreeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them\ncame and looted us, took over what was ours. Yet we have not done this to any other nation. We\nhave not conquered anyone. We have not grabbed their land, their culture and their history and\ntried to enforce our way of life on them. Why? Because we respect the freedom of others. That\nis why my FIRST VISION is that of FREEDOM. I believe that India got its first vision of this in\n1857, when we started the war of Independence. It is this freedom that we must protect and\nnurture and build on. If we are not free, no one will respect us.\nW

## MapReduce to Summerize Large Documents

In [9]:
from langchain.text_splitter import RecursiveCharacterTextSplitter
docs

[Document(metadata={'producer': 'Skia/PDF m127', 'creator': 'Chromium', 'creationdate': '2025-10-27T13:45:14+00:00', 'moddate': '2025-10-27T13:45:14+00:00', 'source': 'drapj.pdf', 'total_pages': 2, 'page': 0, 'page_label': '1'}, page_content='Here is the full text of Dr. APJ Abdul Kalam\'s "My Vision for India" speech, delivered at IIT\nHyderabad on 25 May 2011, without any summarization:\nI have three visions for India. In 3000 years of our history people from all over the world have\ncome and invaded us, captured our lands, conquered our minds. From Alexander onwards the\nGreeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them\ncame and looted us, took over what was ours. Yet we have not done this to any other nation. We\nhave not conquered anyone. We have not grabbed their land, their culture and their history and\ntried to enforce our way of life on them. Why? Because we respect the freedom of others. That is\nwhy my FIRST VISION is that of FR

In [10]:
final_documents = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=100).split_documents(docs)
final_documents

[Document(metadata={'producer': 'Skia/PDF m127', 'creator': 'Chromium', 'creationdate': '2025-10-27T13:45:14+00:00', 'moddate': '2025-10-27T13:45:14+00:00', 'source': 'drapj.pdf', 'total_pages': 2, 'page': 0, 'page_label': '1'}, page_content='Here is the full text of Dr. APJ Abdul Kalam\'s "My Vision for India" speech, delivered at IIT\nHyderabad on 25 May 2011, without any summarization:\nI have three visions for India. In 3000 years of our history people from all over the world have\ncome and invaded us, captured our lands, conquered our minds. From Alexander onwards the\nGreeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them\ncame and looted us, took over what was ours. Yet we have not done this to any other nation. We\nhave not conquered anyone. We have not grabbed their land, their culture and their history and\ntried to enforce our way of life on them. Why? Because we respect the freedom of others. That is\nwhy my FIRST VISION is that of FR

In [11]:
# define the prompt template for text summerization using map_reduce chain
from langchain.prompts import PromptTemplate
prompt_template = """
Write short summary of the following speech in concise points,
speech: {text}
"""
map_prompt = PromptTemplate(
    input_variables=["text"],
    template=prompt_template
)

In [12]:
# final prompt for text summerization using map_reduce chain
final_prompt = """
Provide the final summary based on the following important points:
Add a motivation title , start the precise summery with an introduction and provide the summery
in number of points for the speech.
speech: {text}
"""
final_prompt=PromptTemplate(
    input_variables=["text"],
    template=final_prompt
)

In [13]:
# summery chain using map_reduce
from langchain.chains.summarize import load_summarize_chain
chain = load_summarize_chain(
    llm, 
    chain_type="map_reduce", 
    map_prompt=map_prompt, 
    combine_prompt=final_prompt, 
    verbose=True
)
summary = chain.run(final_documents)
summary



> Entering new MapReduceDocumentsChain chain...


> Entering new LLMChain chain...
Prompt after formatting:

Write short summary of the following speech in concise points,
speech: Here is the full text of Dr. APJ Abdul Kalam's "My Vision for India" speech, delivered at IIT
Hyderabad on 25 May 2011, without any summarization:
I have three visions for India. In 3000 years of our history people from all over the world have
come and invaded us, captured our lands, conquered our minds. From Alexander onwards the
Greeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them
came and looted us, took over what was ours. Yet we have not done this to any other nation. We
have not conquered anyone. We have not grabbed their land, their culture and their history and
tried to enforce our way of life on them. Why? Because we respect the freedom of others. That is
why my FIRST VISION is that of FREEDOM. I believe that India got its first vision of this in 1857,
when

Token indices sequence length is longer than the specified maximum sequence length for this model (1513 > 1024). Running this sequence through the model will result in indexing errors




> Entering new LLMChain chain...
Prompt after formatting:

Provide the final summary based on the following important points:
Add a motivation title , start the precise summery with an introduction and provide the summery
in number of points for the speech.
speech: Here is a concise summary of Dr. APJ Abdul Kalam's "My Vision for India" speech in 3 points:

* **Vision 1: Freedom**: India must protect and nurture its freedom, which is rooted in respecting the freedom of others, and was first envisioned during the 1857 War of Independence.
* **Vision 2: Development**: India must see itself as a developed nation, self-reliant and self-assured, building on its 10% growth rate, falling poverty levels, and global recognition of its achievements.
* **Vision 3: Strength**: India must stand up to the world as a strong military and economic power, as only strength respects strength, to earn the respect of other nations and become a leading global player.

"My Vision for India" speech: 
India c

'**"Empowering a Brighter Tomorrow: Uniting for a Stronger India"**\n\nIntroduction:\nDr. APJ Abdul Kalam\'s "My Vision for India" speech and the subsequent speech by another Indian leader emphasize the need for India to become a developed, powerful, and prosperous nation. The speeches highlight the importance of teamwork, self-reliance, and innovation in achieving this vision. The key points from both speeches can be summarized as follows:\n\nHere is a concise summary of the speeches in 6 points:\n\n1. **Protecting Freedom**: India must protect and nurture its freedom, which is rooted in respecting the freedom of others, and was first envisioned during the 1857 War of Independence.\n2. **Becoming a Developed Nation**: India can become a developed nation by 2020, leveraging its advantages in population, technology, human resources, and natural resources to become the most powerful and prosperous nation in the world.\n3. **Overcoming Obstacles**: Despite progress in various fields, Indi

# RefineChain Text Summerization

In [14]:
# RefineChain Text Summerization

chain = load_summarize_chain(
    llm,
    chain_type="refine",
    verbose=True
)
summary = chain.run(final_documents)
summary



> Entering new RefineDocumentsChain chain...


> Entering new LLMChain chain...
Prompt after formatting:
Write a concise summary of the following:


"Here is the full text of Dr. APJ Abdul Kalam's "My Vision for India" speech, delivered at IIT
Hyderabad on 25 May 2011, without any summarization:
I have three visions for India. In 3000 years of our history people from all over the world have
come and invaded us, captured our lands, conquered our minds. From Alexander onwards the
Greeks, the Turks, the Moguls, the Portuguese, the British, the French, the Dutch, all of them
came and looted us, took over what was ours. Yet we have not done this to any other nation. We
have not conquered anyone. We have not grabbed their land, their culture and their history and
tried to enforce our way of life on them. Why? Because we respect the freedom of others. That is
why my FIRST VISION is that of FREEDOM. I believe that India got its first vision of this in 1857,
when we started the war of Indepen

'In the words of Mahatma Gandhi “BE THE CHANGE YOU WISH TO SEE IN THE WORLD.” \n⁂\nIt is not just enough to point fingers at our system, we need to come together to change it for the better. \n⁂\nWe need to be the leaders of today, not tomorrow. We need to take ownership of our problems and find solutions. \n⁂\nAs a nation, we have all the potential to be a superpower, but it requires the collective effort of all of us. \n⁂\nMy vision for India is not just a dream, but a reality that we can achieve if we work together towards it.\n⁂\nWe can learn from the experiences of the Western countries and apply them to our own context to make India a better place. \n⁂\nIt is time for us to take charge and be the change we wish to see. \n⁂\nLet us join hands and work together to make our country a superpower.\n⁂\nIn conclusion, my vision for India is a nation where every citizen has access to quality education, healthcare, and opportunities for growth and development. \n⁂\nA nation where every ci